posix/syscall.c

```diff
@@ -0,0 +1,598 @@
+            const struct oe_iovec* iov = (const struct oe_iovec*)arg2;
+            int iovcnt = (int)arg3;
+
+            ret = oe_writev(fd, iov, iovcnt);
+            goto done;
+        }
+        case OE_SYS_read:
+        {
+            int fd = (int)arg1;
+            void* buf = (void*)arg2;
+            size_t count = (size_t)arg3;
+
+            ret = oe_read(fd, buf, count);
+            goto done;
+        }
+        case OE_SYS_write:
+        {
+            int fd = (int)arg1;
+            const void* buf = (void*)arg2;
+            size_t count = (size_t)arg3;
+
+            ret = oe_write(fd, buf, count);
+            goto done;
+        }
+        case OE_SYS_close:
+        {
+            int fd = (int)arg1;
+
+            ret = oe_close(fd);
+            goto done;
+        }
+        case OE_SYS_dup:
+        {
+            int fd = (int)arg1;
+
+            ret = oe_dup(fd);
+            goto done;
+        }
+        case OE_SYS_dup2:
+        {
+            int fd = (int)arg1;
+            int newfd = (int)arg2;
+
+            ret = oe_dup2(fd, newfd);
+            goto done;
+        }
+        case OE_SYS_stat:
+        {
+            const char* pathname = (const char*)arg1;
+            struct oe_stat* buf_out = (struct oe_stat*)arg2;
+            struct oe_stat buf;
+
+            ret = oe_stat(pathname, &buf);
+            memcpy(buf_out, &buf, sizeof(struct oe_stat));
+            goto done;
+        }
+        case OE_SYS_link:
+        {
+            const char* oldpath = (const char*)arg1;
+            const char* newpath = (const char*)arg2;
+
+            ret = oe_link(oldpath, newpath);
+            goto done;
+        }
+        case OE_SYS_unlink:
+        {
+            const char* pathname = (const char*)arg1;
+
+            ret = oe_unlink(pathname);
+            goto done;
+        }
+        case OE_SYS_rename:
+        {
+            const char* oldpath = (const char*)arg1;
+            const char* newpath = (const char*)arg2;
+
+            ret = oe_rename(oldpath, newpath);
+            goto done;
+        }
+        case OE_SYS_truncate:
+        {
+            const char* path = (const char*)arg1;
+            ssize_t length = (ssize_t)arg2;
+
+            ret = oe_truncate(path, length);
+            goto done;
+        }
+        case OE_SYS_mkdir:
+        {
+            const char* pathname = (const char*)arg1;
+            uint32_t mode = (uint32_t)arg2;
+
+            ret = oe_mkdir(pathname, mode);
+            goto done;
+        }
+        case OE_SYS_rmdir:
+        {
+            const char* pathname = (const char*)arg1;
+            ret = oe_rmdir(pathname);
+            goto done;
+        }
+        case OE_SYS_access:
+        {
+            const char* pathname = (const char*)arg1;
+            int mode = (int)arg2;
+
+            ret = oe_access(pathname, mode);
+            goto done;
+        }
+        case OE_SYS_getdents:
+        case OE_SYS_getdents64:
```

Comment:
To me it seems like all this code here can be made optional by using the syscall hooks approach that we used in some tests.

```c

/* pseudo code illustration */
typedef long(*oe_syscall_hook_t)(int n, oe_va_list ap);

oe_syscall_hook_t _hooks[OE_MAX_SYSCALL_HOOKS]; // corresponds to maximum syscall number + 1

long syscall(long number, ...)
{
    oe_va_list ap;
    oe_va_start(ap, number);

    if (_hooks[number] == NULL)
         OE_RAISE(OE_UNSUPPORTED_SYSCALL);

     return _hooks(number, ap);
}
```

The functions that register the devices can also register the syscall hooks. 
Any reason we gave up on the syscall hooks idea from before?



CC: @anitagov @CodeMonkeyLeet

Reply:
I think you might be getting **posix/syscall.c** confused with **libc/sgx/syscall.c** (which still contains the hook framework I added before). We are not abandoning the hook framework. Rather **oe_syscall()** is a hook and you will see later how it is called from **libc/sgx/syscall.c** when the **libc** changes PR is ready. Maybe **oe_posix_syscall()** would be a better name.

Reply:
@mikbras I looked searched for syscall in posix standard. There does not seem to be a match. Do you happen to have a posix reference handy?

The [linux reference](http://man7.org/linux/man-pages/man2/syscall.2.html) says:
> syscall() saves CPU registers before making the system call, restores
       the registers upon return from the system call, and stores any error
       code returned by the system call in errno(3) if an error occurs.

Do you know which registers the syscall library function is supposed to preserve?

Reply:
Since we are not doing an actual syscall (we are calling a function called syscall() instead), I am not sure why you are concerned about register restoration.

Reply:
Maybe I am misunderstanding the design. 
oe_syscall implements GLIBC (or BSD?) syscall() library function. Yes?

If no, then you can ignore my comment.

If yes, wewant to make sure if the registers are correctly preserved described by the linux reference.
Say, for example, does syscall library function make any guarantees of floating point control word?
The normal syscall library function would invoke a system call instruction (or equivalent) at some point.
In our case, behind the scenes we may end up making an ocall that could change the FPU state. 
Or say one of the device implementation in the enclave side could call some floating point implementation that can change the FPU state. In that case, would our syscall library function violate the BSD/glibc syscall function guarantees?

Reply:
Ah! You found  an existing bug in **syscall()** defined in **libc**. I can fix that bug when the **libc** changes are ready for review. I need to write an assembly function.

But note that the **oe_syscall()** is a hook in this PR is called by **syscall()** from **libc**. This function is fine as it is.

Reply:
Very interesting thread.